### PR TITLE
ts: Add program.coder.types for encoding/decoding user-defined types

### LINKED ISSUE #1930: Expose coder for user defined types as `program.coder.types`.
I've run into a situation recently when I needed to manually deserialize a slice of account data into a type that I have in my IDL, but which is not an account type. Here's like my Rust part looks like:
```rust
#[account]
pub struct Mintlist {
    pub mints: Vec<MintInfo>
}

#[derive(AnchorSerialize, AnchorDeserialize, Clone, PartialEq)]
pub struct MintInfo {
    pub minted: bool,           // 1
    pub metadata_url: [u8; 96], // 96
}
```
On the client I would like to use something like this:

```ts
const mintInfo = program.coder.types.decode("MintInfo", dataSlice)
```

Looking at the codebase it seems like a trivial thing to add and I'm happy to submit a PR if you think it makes sense.

### PR DESCRIPTION
**Pull request by vovacodes**
fixes https://github.com/project-serum/anchor/issues/1930

### CODE REVIEW

File: ts/src/coder/index.ts [hunk 1]
```diff
@@ -28,6 +28,11 @@ export interface Coder {
    * Coder for events.
    */
   readonly events: EventCoder;
+
+  /**
+   * Coder for user-defined types.
+   */
+  readonly types: TypesCoder<T>;
```
Comment (armaniferrante) [inline ts/src/coder/index.ts:35]: Please add a CHANGELOG.md entry for this change.

Comment (vovacodes) [reply ts/src/coder/index.ts:35]: Done

**Comment by armaniferrante** [comment]
Thank you!

File: ts/src/coder/borsh/types.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  private idl: Idl;
+
+  public constructor(idl: Idl) {
+    if (idl.types === undefined) {
+      this.typeLayouts = new Map();
+      return;
+    }
+    const layouts: [N, Layout][] = idl.types.map((acc) => {
+      return [acc.name as N, IdlCoder.typeDefLayout(acc, idl.types)];
+    });
```
[the remark this answers is not in the document]
Comment (snawaz) [reply ts/src/coder/borsh/types.ts:28]: OR, we have a problem is `event.js` as it does not use `accounts`. Events could use accounts as well(especially the smaller ones).

**Pull request merged**